CAMEL-20404: fixes

orpiske opened the pull request.

orpiske committed:
CAMEL-20404: fix incorrect handling of null bodies

orpiske commented:
@lburgazzoli  If this one is OK, then I'll send a fix for the other operations in a separate PR. It depends on #4.

lburgazzoli commented on components/camel-qdrant/src/main/java/org/apache/camel/component/qdrant/QdrantProducer.java:
```diff
@@ -235,6 +241,26 @@ private boolean delete(Exchange exchange, AsyncCallback callback) {
+        final Message in = exchange.getMessage();
+        final VectorParams body = in.getBody(VectorParams.class);
+
+        ObjectHelper.notNull(body, "A payload of type VectorParams.class is expected");
```
I tend to use the `QuadrantActionException` to include also the action that was caused the exception in the message

orpiske replied:
In general, we have been using the `ObjectHelper.notNull` to handle cases like this. It currently does not support an specific exception like that, but I think I can make a modification to it if we want to use a custom exception.

orpiske replied:
I'll propose a PR to core so we can use it like this:

```
ObjectHelper.notNull(body, "A payload of type VectorParams.class is expected", () -> new QdrantActionException(
                    QdrantAction.DELETE,
                    "A payload of type PointsSelector, PointId or Filter is expected"));
```

lburgazzoli replied:
fine with me either way.

I tend to use some better/custom exception when a failure happen during routing as it is easier to spot the reason/context as opposed of throwing a generic one

orpiske replied:
Alright. So, let's use the new feature so we can keep the code base consistent, but also provide the contextual information as provided by the exception. I have a PR open there. I'll merge it soon.

lburgazzoli replied:
you may need to rebase since I merged the other pr

lburgazzoli replied:
Maybe we should try to avoid duplicating the message ? so something like  

```java 
ObjectHelper.notNull(
    body,
     () -> new QdrantActionException(
    	QdrantAction.DELETE,
        "A payload of type PointsSelector, PointId or Filter is expected"));
```

but I don't know

valdar replied:
I have always thought that an `exchange.setException(...)` was the idiomatic way of rising exceptions the camel way in async exchange processing... Most probably at the end of the day both ways result in the same behavior but still I find `exchange.setException(...)` more "camel idiomatic". 

My 2c.

orpiske replied:
Actually the behavior is not the same. Using just `exchange.setException` (even followed by a return true/false) causes the code to crash/hang upon receiving a null body (it seems not to force the exchange to stop processing).

Recently, we have consolidated in using the `ObjectHelper.notNull` as the way to check for required parameters/body as it keeps the behavior and code base consistent.

lburgazzoli replied:
this is a mistake as the callback is not invoked when setting an exception on the exchange (which I'm fixing now)

orpiske replied:
The root of the problem might be that we are checking for requirements a bit too late (IMHO, we should check this before `process`) and we should leave the `setException` for processing exceptions.

orpiske replied:
@lburgazzoli also, let's split the parameter checking from the processing exceptions, so it's consistent with the rest of the components. After you implement this fix, I'll adjust the code to move the parameter checking.

lburgazzoli replied:
I'm not sure I follow how that should be spitted to be honest

orpiske replied:
> I'm not sure I follow how that should be spitted to be honest

I meant that we should be checking for the body, headers, etc before we actually get to do the processing (`createCollection`, `retrieve`, etc). But that's not necessary after you've changed the code to use the `try/catch` as @davsclaus suggested.

orpiske committed:
CAMEL-20404: increase test coverage

- Added unit tests
- Split unit and integration tests

davsclaus commented:
when the process method calls 4 different operations, then its better practice to have it do it via 

try catch and let sub method throw exception to ensure correct behaviour
in catch exception you set exception on exchange and invoke callback done true, and return true

then all these 4 sub methods can just return false and invoke the callback when they have an answer and are ready to continue

the current code has a few mistakes forgetting to invoke callback done(true) when an exception was thrown etc

lburgazzoli commented:
> the current code has a few mistakes forgetting to invoke callback done(true) when an exception was thrown etc

can you check it again ? I mean, in this PR https://github.com/apache/camel/pull/13109

davsclaus commented:
You can use getMandatoryBody(xxx) if the body MUST be present and of a given type, that gives a better error for the end user

Those Object notNull checks are usually for wiring of the classes during bootstrap etc.
For data processing then we have specialized exception for that - NoSuchHeader, InvalidBody...

orpiske commented:
I updated the tests to accommodate the latest changes, split unit from integration tests, added tests checking for required parameters and adjusted the header check to use the `NoSuchHeaderException` instead of the `ObjectHelper.notNull`.

orpiske commented:
> You can use getMandatoryBody(xxx) if the body MUST be present and of a given type, that gives a better error for the end user

One challenge here is that we can only check for either `nullability` or a single type (i.e.; `getMandatoryBody(Points.PointStruct.class)`. We could try to implementer a converter, though.

davsclaus commented:
mandatory body will convert to the given type if the body is not already of that

orpiske commented:
> mandatory body will convert to the given type if the body is not already of that

Yeah, but what I am saying is that there's more than one possible type to be converted to.

orpiske commented:
> Yeah, but what I am saying is that there's more than one possible type to be converted to.

@davsclaus @lburgazzoli this is what I meant: https://github.com/orpiske/camel/commit/0632180303b5b3c65cec3fd39a19cdb70e662d7a

This way we can use the `getMandatoryBody` as Claus suggested originally. 

if you are OK with that, I'll send a separate PR later ... as we have gone off topic a lot here and driving Luca crazy 😆

lburgazzoli commented:
> > Yeah, but what I am saying is that there's more than one possible type to be converted to.
> 
> 
> 
> @davsclaus @lburgazzoli this is what I meant: https://github.com/orpiske/camel/commit/0632180303b5b3c65cec3fd39a19cdb70e662d7a
> 
> 
> 
> This way we can use the `getMandatoryBody` as Claus suggested originally. 
> 
> 
> 
> if you are OK with that, I'll send a separate PR later ... as we have gone off topic a lot here and driving Luca crazy 😆 

it looks good

orpiske commented:
@lburgazzoli awesome. So, once you merge this one, I'll send a PR with those type converters.
